From 46d7eec602c7cb9bee857cdf1d55c55eeaccb97d Mon Sep 17 00:00:00 2001
From: Timple Raj M <timple@qti.qualcomm.com>
Date: Mon, 13 Apr 2026 19:58:00 +0530
Subject: [PATCH] serial: msm_geni: Set RX FIFO watermark

Program the RX FIFO watermark to 2 during RX start and
setup to ensure reliable RX watermark interrupts.

The value is aligned with the RX FIFO watermark used in
Linux v6.6 drivers/tty/serial/qcom_geni_serial.c to ensure
reliable RX watermark interrupts.

Change-Id: I83db862a19e583f82df522c8c88f8ac3afb65595
Signed-off-by: Timple Raj M <timple@qti.qualcomm.com>
---
 drivers/serial/serial_msm_geni.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/drivers/serial/serial_msm_geni.c b/drivers/serial/serial_msm_geni.c
index 2767ab21a14..f5431ffce88 100644
--- a/drivers/serial/serial_msm_geni.c
+++ b/drivers/serial/serial_msm_geni.c
@@ -55,6 +55,7 @@
 #define SE_UART_RX_PARITY_CFG	0x2a8
 
 #define DEF_TX_WM	2
+#define DEF_RX_WM	2
 
 #define UART_START_READ	0x1
 
@@ -367,6 +368,7 @@ static void qcom_geni_serial_start_rx(struct udevice *dev)
 	status = readl(priv->base + SE_GENI_STATUS);
 
 	geni_se_setup_s_cmd(priv->base, UART_START_READ, 0);
+	writel(DEF_RX_WM, priv->base + SE_GENI_RX_WATERMARK_REG);
 
 	setbits_le32(priv->base + SE_GENI_S_IRQ_EN, S_RX_FIFO_WATERMARK_EN | S_RX_FIFO_LAST_EN);
 	setbits_le32(priv->base + SE_GENI_M_IRQ_EN, M_RX_FIFO_WATERMARK_EN | M_RX_FIFO_LAST_EN);
@@ -395,6 +397,7 @@ static void msm_geni_serial_setup_rx(struct udevice *dev)
 	writel(UART_PACKING_CFG1, priv->base + SE_GENI_RX_PACKING_CFG1);
 
 	geni_se_setup_s_cmd(priv->base, UART_START_READ, 0);
+	writel(DEF_RX_WM, priv->base + SE_GENI_RX_WATERMARK_REG);
 
 	setbits_le32(priv->base + SE_GENI_S_IRQ_EN, S_RX_FIFO_WATERMARK_EN | S_RX_FIFO_LAST_EN);
 	setbits_le32(priv->base + SE_GENI_M_IRQ_EN, M_RX_FIFO_WATERMARK_EN | M_RX_FIFO_LAST_EN);
@@ -640,6 +643,7 @@ static inline void _debug_uart_init(void)
 	phys_addr_t base = CONFIG_VAL(DEBUG_UART_BASE);
 
 	geni_serial_init(&init_dev);
+	writel(DEF_RX_WM, priv->base + SE_GENI_RX_WATERMARK_REG);
 	geni_serial_baud(base, CLK_DIV, CONFIG_BAUDRATE);
 	qcom_geni_serial_start_tx(base);
 }
-- 
2.34.1

